File: src/bunnymark.js
// title: Bunnymark in JavaScript
// author: Rob Loach
// desc: Benchmarking tool to see how many bunnies can fly around the screen, using JavaScript.
// input: gamepad
// script: js

const screenWidth = 240
const screenHeight = 136
const toolbarHeight = 6

function GetRandomValue(min, max) {
  return Math.random() * (max - min) + min
}

function Bunny() {
	this.x = GetRandomValue(0, screenWidth - Bunny.width)
	this.y = GetRandomValue(0, screenHeight - Bunny.height)
	this.speedX = GetRandomValue(-100, 100) / 60
	this.speedY = GetRandomValue(-100, 100) / 60
	this.sprite = 1

	this.draw = function() {
		spr(this.sprite, this.x, this.y, 1, 1, 0, 0, 4, 4)
	}

	this.update = function() {
		this.x += this.speedX
		this.y += this.speedY

		if (this.x + Bunny.width > screenWidth) {
			this.x = screenWidth - Bunny.width
			this.speedX *= -1
		}
		if (this.x < 0) {
			this.x = 0
			this.speedX *= -1
		}
		if (this.y + Bunny.height > screenHeight) {
			this.y = screenHeight - Bunny.height
			this.speedY *= -1
		}

		if (this.y < toolbarHeight) {
			this.y = toolbarHeight
			this.speedY *= -1
		}
	}
}

Bunny.width = 26
Bunny.height = 32

function FPS() {
	this.value = 0
	this.frames = 0
	this.lastTime = -1000
	this.getValue = function() {
		if (time() - this.lastTime <= 1000) {
			this.frames = this.frames + 1
		}
		else {
			this.value = this.frames
			this.frames = 0
			this.lastTime = time()
		}
		return this.value
	}
}

const fps = new FPS()
const bunnies = []
bunnies.push(new Bunny())

function TIC() {
	// Input
	if (btn(0)) {
		for (var i = 0; i < 5; i++) {
			bunnies.push(new Bunny())
		}
	}
	else if (btn(1)) {
		for (var i = 0; i < 5; i++) {
			bunnies.pop()
		}
	}

	// Update
	for (var i in bunnies) {
		bunnies[i].update()
	}

	// Draw
	cls(15)
	for (var i in bunnies) {
		bunnies[i].draw()
	}
	rect(0, 0, screenWidth, toolbarHeight, 0)
	print("Bunnies: " + bunnies.length, 1, 0, 11, false, 1, false)
	print("FPS: " + fps.getValue(), screenWidth / 2, 0, 11, false, 1, false)
}

File: src/bunnymark.lua
-- title: Bunnymark in Lua
-- author: @RobLoach and Rabia Alhaffar
-- desc: Benchmarking tool to see how many bunnies can fly around the screen, using Lua.
-- input: gamepad
-- script: lua
-- version: 1.2.0

local screenWidth = 240
local screenHeight = 136
local toolbarHeight = 6
local t = 0

local Bunny
do
    local _class_0
    local _base_0 = {
        draw = function(self)
            return spr(self.sprite, self.x, self.y, 1, 1, 0, 0, 4, 4)
        end,
        update = function(self)
            self.x = self.x + self.speedX
            self.y = self.y + self.speedY
            if self.x + self.__class.width > screenWidth then
                self.x = screenWidth - self.__class.width
                self.speedX = self.speedX * -1
            end
            if self.x < 0 then
                self.x = 0
                self.speedX = self.speedX * -1
            end
            if self.y + self.__class.height > screenHeight then
                self.y = screenHeight - self.__class.height
                self.speedY = self.speedY * -1
            end
            if self.y < toolbarHeight then
                self.y = toolbarHeight
                self.speedY = self.speedY * -1
            end
        end
    }
    _base_0.__index = _base_0
    _class_0 = setmetatable({
        __init = function(self)
            self.x = math.random(0, screenWidth - self.__class.width)
            self.y = math.random(0, screenHeight - self.__class.height)
            self.speedX = math.random(-100, 100) / 60
            self.speedY = math.random(-100, 100) / 60
            self.sprite = 1
        end,
        __base = _base_0,
        __name = "Bunny"
    }, {
        __index = _base_0,
        __call = function(cls, ...)
            local _self_0 = setmetatable({}, _base_0)
            cls.__init(_self_0, ...)
            return _self_0
        end
    })
    _base_0.__class = _class_0
    local self = _class_0
    self.width = 26
    self.height = 32
    Bunny = _class_0
end

local FPS
do
    local _class_0
    local _base_0 = {
        getValue = function(self)
            if time() - self.lastTime <= 1000 then
                self.frames = self.frames + 1
            else
                self.value = self.frames
                self.frames = 0
                self.lastTime = time()
            end
            return self.value
        end
    }
    _base_0.__index = _base_0
    _class_0 = setmetatable({
        __init = function(self)
            self.value = 0
            self.frames = 0
            self.lastTime = 0
        end,
        __base = _base_0,
        __name = "FPS"
    }, {
        __index = _base_0,
        __call = function(cls, ...)
            local _self_0 = setmetatable({}, _base_0)
            cls.__init(_self_0, ...)
            return _self_0
        end
    })
    _base_0.__class = _class_0
    FPS = _class_0
end

local fps = FPS()
local bunnies = { }
table.insert(bunnies, Bunny())

TIC = function()
    if t == 0 then
        music(0)
    end
    if t == 6 * 64 * 2.375 then
        music(1)
    end

    t = t + 1
    if btn(0) then
        for i = 1, 5 do
            table.insert(bunnies, Bunny())
        end
    elseif btn(1) then
        for i = 1, 5 do
            table.remove(bunnies, 1)
        end
    end

    for i, item in pairs(bunnies) do
        item:update()
    end

    cls(15)
    for i, item in pairs(bunnies) do
        item:draw()
    end

    rect(0, 0, screenWidth, toolbarHeight, 0)
    print("Bunnies: " .. #bunnies, 1, 0, 11, false, 1, false)
    return print("FPS: " .. fps:getValue(), screenWidth / 2, 0, 11, false, 1, false)
end
